src/send/mod.rs

```diff
@@ -0,0 +1,99 @@
+#[cfg(test)]
+mod test;
+
+use rocket::{
+  data::{self, FromData},
+  http::Status,
+  Data,
+  Outcome,
+  Request,
+};
+use rocket_contrib::{Json, Value};
+use std::{borrow::Cow, collections::HashMap};
+use validator::{self, ValidationError};
+
+#[derive(Debug, Deserialize)]
+struct Body
+{
+  text: String,
+  html: Option<String>,
+}
+
+#[derive(Debug, Deserialize)]
+struct Email
+{
+  to: Vec<String>,
+  cc: Option<Vec<String>>,
+  subject: String,
+  body: Body,
+}
+
+impl FromData for Email
+{
+  type Error = ValidationError;
+
+  fn from_data(request: &Request, data: Data) -> data::Outcome<Self, Self::Error>
+  {
+    let result = Json::<Email>::from_data(request, data);
+    match result
+    {
+      Outcome::Success(json) =>
+      {
+        let email = json.into_inner();
+
+        if email.to.len() == 0
+        {
+          return fail();
+        }
+
+        if let Some(outcome) = validate_addresses(&email.to)
+        {
+          return outcome;
```

Comment:
When I first read this I was a little confused, become `outcome` sounds a bit like is succeeded.  Would it read better as `Some(error)` or similar?

Reply:
Tbh, all the naming here dissatisfies me. `validate_addresses` is misleading too, given the implementation. I think I'm better at naming things in the a.m., will sleep on it! 😄